#### Jérémy TREMBLAY

# TD1: Decision Tree

In [1]:
import math

## Task 1: Build decision tree

| Force vent (FV) | Présence précipitations (PP) | État piste (EP) | Décision |
|-----------------|-----------------------------|-----------------|----------|
| 1               | Faible                      | Faux                          | Correcte  | oui     |
| 2               | Faible                      | Faux                          | Dégradée | oui     |
| 3               | Faible                      | Vrai                          | Correcte  | oui     |
| 4               | Modérée                     | Faux                          | Correcte  | oui     |
| 5               | Modérée                     | Vrai                          | Dégradée | non     |
| 6               | Modérée                     | Vrai                          | Correcte  | non     |
| 7               | Dangereuse                  | Faux                          | Correcte  | non     |
| 8               | Dangereuse                  | Vrai                          | Dégradée | non     |


**Consigne :** A partir de ces données, construisez un arbre de décision pure.

First we need to normalize data. Here is the new board.

| Vent faible (VF) | Vent modéré (VM) | Vent violent (VV) | Présence précipitations (PP)   | État piste (EP)   | Décision   |
|:------------:|:-------------:|:----------:|:-----------------------------|:-------------------|:-----------|
|    vrai    |    faux    |   faux   | Faux                        | Correcte          | oui        |
|    vrai    |    faux    |   faux   | Faux                        | Dégradée         | oui        |
|    vrai    |    faux    |   faux   | Vrai                        | Correcte          | oui        |
|    faux    |    vrai    |   faux   | Faux                        | Correcte          | oui        |
|    faux    |    vrai    |   faux   | Vrai                        | Dégradée         | non        |
|    faux    |    vrai    |   faux   | Vrai                        | Correcte          | non        |
|    faux    |    faux    |   vrai   | Faux                        | Correcte          | non        |
|    faux    |    faux    |   vrai   | Vrai                        | Dégradée         | non        |


We first compute each entropie for each criteria value. 

In [12]:
print("J(VF = VRAI) = -3/3 * log2(3/3) = {:.2f}".format(-3/3 * math.log2(3/3)))
print("J(VF = FAUX) = -1/5 * log2(1/5) - 4/5 * log2(4/5) = {:.2f}".format(-1/5 * math.log2(1/5) - 4/5 * math.log2(4/5)))
print("J(VM = VRAI) = -1/3 * log2(1/3) - 2/3 * log2(2/3) = {:.2f}".format(-1/3 * math.log2(1/3) - 2/3 * math.log2(2/3)))
print("J(VM = FAUX) = -3/5 * log2(3/5) - 2/5 * log2(2/5) = {:.2f}".format(-3/5 * math.log2(3/5) - 2/5 * math.log2(2/5)))
print("J(VV = VRAI) = -2/2 * log2(2/2) = {:.2f}".format(-2/2 * math.log2(2/2)))
print("J(VV = FAUX) = -4/6 * log2(4/6) - 2/6 * log2(2/6) = {:.2f}".format(-4/6 * math.log2(4/6) - 2/6 * math.log2(2/6)))
print("J(PP = VRAI) = -1/4 * log2(1/4) - 3/4 * log2(3/4) = {:.2f}".format(-1/4 * math.log2(1/4) - 3/4 * math.log2(3/4)))
print("J(PP = FAUX) = -3/4 * log2(3/4) - 1/4 * log2(1/4) = {:.2f}".format(-3/4 * math.log2(3/4) - 1/4 * math.log2(1/4)))
print("J(EP = Correcte) = -3/5 * log2(3/5) - 2/5 * log2(2/5) = {:.2f}".format(-3/5 * math.log2(3/5) - 2/5 * math.log2(2/5)))
print("J(EP = Dégradée) = -1/3 * log2(1/3) - 2/3 * log2(2/3) = {:.2f}".format(-1/3 * math.log2(1/3) - 2/3 * math.log2(2/3)))

J(VF = VRAI) = -3/3 * log2(3/3) = -0.00
J(VF = FAUX) = -1/5 * log2(1/5) - 4/5 * log2(4/5) = 0.72
J(VM = VRAI) = -1/3 * log2(1/3) - 2/3 * log2(2/3) = 0.92
J(VM = FAUX) = -3/5 * log2(3/5) - 2/5 * log2(2/5) = 0.97
J(VV = VRAI) = -2/2 * log2(2/2) = -0.00
J(VV = FAUX) = -4/6 * log2(4/6) - 2/6 * log2(2/6) = 0.92
J(PP = VRAI) = -1/4 * log2(1/4) - 3/4 * log2(3/4) = 0.81
J(PP = FAUX) = -3/4 * log2(3/4) - 1/4 * log2(1/4) = 0.81
J(EP = Correcte) = -3/5 * log2(3/5) - 2/5 * log2(2/5) = 0.97
J(EP = Dégradée) = -1/3 * log2(1/3) - 2/3 * log2(2/3) = 0.92


We then compute H for each criteria.  

H(oui, VV) = 2/8 J(VV = VRAI) + 6/8 J(VV = FAUX) = 0.81

Then we can create new boards.

Then we follow the process again and again and again...